function_wrapper_test.go

```diff
@@ -0,0 +1,56 @@
+package gosnowflake
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"testing"
+)
+
+var (
+	goWrapperCalled chan struct{}
+	once       sync.Once
+)
+
+// Initialize the goWrapperCalled channel which we will use for testing the goroutine wrap function
+func initGoWrapperCalledChannel() {
+	once.Do(func() {
+		goWrapperCalled = make(chan struct{})
+	})
+}
+
+// Check if the goWrapperCalled by making sure channel is clsoed
+func isGoWrapperChannelClosed() bool {
+	select {
+	case <-goWrapperCalled:
+		return true
+	default:
+		return false
+	}
+}
+
+// this is the go wrapper function we are going to pass into GoroutineWrapper.
+// we will know that this has been called if the channel is closed
+var closeGoWrapperCalledChannel = func(f func()) {
+	if !isGoWrapperChannelClosed() {
+		close(goWrapperCalled)
+	}
+	
+	f()
+}
+
+func TestGoWrapper(t *testing.T) {
+	runDBTest(t, func(dbt *DBTest) {
+		initGoWrapperCalledChannel()
+		GoroutineWrapper = closeGoWrapperCalledChannel
+
+		numrows := 100000
+		withCancelCtx, cancel := context.WithCancel(context.Background())
+		// using async mode because I know that will trigger a goroutine to be fired off
```

Comment:
I'm curious, what happens here? :)

Reply:
Like why am I using `WithAsyncMode` to kick off a goroutine rather than just doing it in this function? I wanted to demonstrate that the changes I made in the code would really wrap the goroutine like I intended, and I knew that using this flag would trigger the goroutine on line ~40 of async.go. Alternatively though, if we just want to test that the wrapper func works like expected when we use it to wrap a go routine we can just do something like this. Whatever you think is preferred 

```
func TestGoWrapper(t *testing.T) {
	t.Cleanup(
		func() {
			// set the global wrapper called variable back to false
			goWrapperCalled = false
		},
	)

	// set the global variable which holds our desired wrapper to the one which switches 
	// goWrapperCalled to true
	GoroutineWrapper = closeGoWrapperCalledChannel

	// define a channel so we will know when our goroutine is done 
	ch := make(chan struct{})

	// Start a goroutine that closes the channel after a delay, wrapped by our wrapper
	go GoroutineWrapper(
		context.Background(),
		func() {
			time.Sleep(100 * time.Millisecond)
			close(ch)
		},
	)

	// wait for our pretend work to be done
	<-ch

	// check that our wrapper was called 
	assertTrueF(t, goWrapperCalled, "channel should be closed, indicating our wrapper worked")
}
```